### DISCUSSION TITLE
Support for cascaded inverters

### DISCUSSION THREAD

**Discussion by Kugelfang666** [wlcrs/huawei_solar#521] [Connecting to the inverter]
### Describe your feature request

I just received my installation with **two** Huawei 10KTL-M1 inverters and a Fusion Solar Wallbox. All items are updated to the newest firmware versions avaliable.

The Master inverter is connected to my local network via the Huawei Smart Dongle. Since I the Wallbox needs to read the SmartMeter via this interface, I cannot use it for this integration (one connection only, see https://github.com/wlcrs/huawei_solar/issues/491)

I did not manage to have the Inverters join my home WiFi with their build in WiFi module for some reason. Hence, each of the two inverters is currently running their own AP. My home assistant is connected to the AP of the master inverter using WiFi. Wit this integration I'm able to connect and read the states of inverter 1, the smart meter and the battery. However, I **do not see the cascaded inverter #2** 

It would be great if cascaded inverted would also be supported by this integration

![Bildschirmfoto 2023-10-29 um 16 44 15](https://github.com/wlcrs/huawei_solar/assets/3289654/37907c06-fe9a-408e-bc04-08b5f0fdede4)

### Proper usage

- [X] I confirm that this is not a bug report or support request
- [X] I confirm that this feature request is within the stated scope of the integration

**Comment by wlcrs** [comment]
I cannot make the slave inverter magically appear as a cascaded inverter if it isn't exposed via the WiFi AP of the primary inverter.

3 possible solutions:

1) Use a modbus proxy to sit between the "Smart Dongle" on HA/Wallbox.
2) Reconfigure the AP of the second inverter so that it doesn't have a conflicting 192.168.200.1 IP-address anymore (I don't know the Huawei settings dialogs by hard, so not entirely sure if this is possible), and then connect HA to that inverters WiFi AP as well.
3) Use a dedicated router running OpenWRT and do some networking wizardry with VLANs, and port forwardings to prevent those two colliding IP-ranges from conflicting with each other.

I'll convert this issue to a discussion, as there is no work for me to be done here. I won't give any extra support on how to realise any of the proposed solutions, but I do hope that you document them in the discussion or on the Wiki for future users.

**Reply by Kugelfang666** [reply]
Thanks a lot @wlcrs for the detailed feedback. I‘ll give the Modbus proxy a try and write up a little how-to in case I manage to get this working. Apparently there is a ready to go addon for HA that i found here: 

https://github.com/Akulatraxas/ha-modbusproxy/

In addition i checked and in fact I CAN adjust the IP range for the inverter AP in the app. If the modbus proxy does not work I‘ll by a second USB-WiFi adapter for my HA, configure the AP in the slave inverter to 192.168.20**1**.X  and try to connect to each with its own AP. 

One more wild Idea that came to my mind: Since both Inverters had a smart Dongle in the box: Do you happen to know if one can attach the second smartdongle to the slave inverter and connect it to my local LAN via Ethernet. If this works I could use the smartdongle of the master inverter for the wallboard and the second dongle on the slave inverter for this integration.

**Reply by Roving-Ronin** [reply]
@Kugelfang666 the USB sdongle MUST be installed in the Primary Inverter, and it will in turn (via the RS485) talk to the Secondary Inverter, thereby allowing utilisation with this integration. You idea of having the dongle in the Secondary and having the secondary talk 'UP' to the Primary inverter will not work.

From what you've said above, it sounds like you have a 2 x 10kW M1 inverters, a USB sDongle (technically 2, as you get one with each inverter BUT you can only use ONE in a cascade, and it MUST be in the primary inverter), a DTSU-666H Smart/Consumption meter and an EV charger (aka Wallbox).

In this setup you should have the :
- DTSU-666H connected to the Primary inverter via RS485
- USB sDongle in the Primary Inverter (with it then connecting to your Home WiFi / Internet  via the WiFi or Ethernet function of the dongle)
- Primary inverter connected to the Secondary inverter via RS485 (cascase)
- EV charger / Wallbox connected via Ethernet to your home network / internet.

As per this diagram (EV charger user manual page 39 - https://photomate.zendesk.com/hc/en-gb/article_attachments/10442149835805 ), but add/imagine the secondary inverter linked via RS485 to the Primary inverter ;-)

![image](https://github.com/wlcrs/huawei_solar/assets/108674933/1b3a8beb-9a8a-436a-b43d-40b48a708740)

Comparing to your logical diagram, it appears you have the RS485 all setup, but you just need to remove the connections to the inverters inbuilt WiFi APs on the 192.168.200.xxx (that are really just for local setup / maintenance using FusionSolar app on mobile) and as WLCRS has suggested, install the modbus proxy in HA, pointing the EV charge  (wallbox) and the WLCRS integration in HA to talk to the Modbus Proxy, and in turn the Modbus Proxy to then be the (one supported modbus tcp connection) to the USB sDongle.

**Reply by Kugelfang666** [reply]
@Roving-Ronin, thanks a lot for looking into this, much appreciated! 
In fact I was suggesting to add the SECOND dongle to the slave inverter. My naive hope was that this way I would have a second modbus server accessible with its own local IP that provides access to the whole set of devices in the bus. 

I‘ll start giving the modbus proxy a try first and report back here. 

Again thanks a lot for this great integration and the discussion!

**Reply by Roving-Ronin** [reply]
@Kugelfang666 sorry just saw this update. Only a single sDongle can be used in a cascade of inverters, and it must be installed into the primary inverter. Many a story of solar installers seeing each inverter comes with a sDongle and plugging one into every inverter in a cascade and the cascade then not working (as only 1 sDongle allowed).

**Reply by Kugelfang666** [reply]
Thanks for looking into this. In fact meanwhile I got everything (mostly) from working. I setup a morbid proxy connected to the dongle and setup this integration to connect to the proxy. Key settings that took me a while to setup correctly were the slave IDs and the correct order of them. Since the integration when setup has a default of 0 I was always trying 0,1,2 assuming that 0 belongs to the dongle when in fact I had to use 1,2

Another issues I was encountering was timeouts on the proxy. I was able to mitigate these by increasing the connection time and timeout setting. 

The only thing I did not manage until now is connecting the Wallbox to the proxy. The Wallbox automatically detects the smartmeter via the dongle so that I cannot manually set it up to talk to the proxy instead.

**Reply by Roving-Ronin** [reply]
@Kugelfang666 where was the proxy timeout setting defined ? Tried to use the proxy myself a while ago, just to allow me to keep options open in future, but found that it was timing out whilst HA and the addons where all starting up.

Have you tried setting the Wallbox to a different network subnet to the sDongle? If the Wallbox has a valid IP (DHCP reserved?) so it geta a valid IP, assuming its limited to broadcasting and detecting the sDongle (and the virtual meter it publishes) only upon the same subnet, this would mean it would fail detecting the sDongle. If this where to ocurr, it would then be interesting to see if it gives the option to manually define an IP Address / Port for the sDongle (or even allows this)?

**Reply by Kugelfang666** [reply]
@Roving-Ronin: I'm using the modus-proxy add-on for home assistant. In the config section of this add-on both timeout and connection_time are exposed as parameters that the user can define. For me connection_time 4s and timeout 20s seems to be working well

**Comment by Kugelfang666** [comment]
Ok so I did some testing today:

First I cut my wallbox from power to free up the TCP-Modbus on the smartdongle. Next I used this integration to connect directly to the dongle to identify the right parameters for the modbus (IP, port, slaveID). I successfully established a connection but also via the smart dongle I only found the master inverter, the battery connected to it and the smart meter. The cascaded inverter was not visible.

Therefore I did not invest further effort in setting up the modbus proxy. Instead I configured the IP for the AP on the second inverter to a non-colliding IP range (192.168.201.1) and added a second USB-WiFi dongle to the HA. 

Basically now my HA instance is connected via Ethernet to my home network and via two dedicated WiFi interfaces each to one of the inverters. This way my wallbox can utilize the dongle while I still see all devices in HA

**Reply by Roving-Ronin** [reply]
@Kugelfang666 

a) If the secondary inverter is correctly wired up using the RS485 ports to link it to the primary inverter, you should be able to 'see' it in WLCRS. Was it visible within FusionSolar, confirming that the secondary would therefore be communicating to the primary, that in turn is connected to FusionSolar ?
If not, wondering if the installer remembered to go into FusionSolar (mobile app) and during the initial configuration of each inverter set each inverter with its own COM Address (i.e. primary = address 1 / secondary = address 2).

b) The Wallbox if using the ethernet connection to the LAN, that then talks to the sDongle on the primary inverter, is using (can't rmemebr the official term) a 'virtual' meter (aka virtual DDSU/DTSU-666) that is provided by the modbus-tcp port on the sDongle. As you've worked out, this consumes the single modbus-tcp connection that the sDongle supports :-(

Long term 'neat' solution will be when the Energy Management Assistant Huawei (EMMA-A01/A02) ships, hopefully this is before the end of 2023 annd I note there has been a bunch of firmware updates, user manuals etc all drop on the Huawei Support site for EMMA / WallBox, in the last 2 weeks. I've already opened an issue (more of a place to start discussing / idea toss etc) for EMMA support in future at: https://github.com/wlcrs/huawei_solar/issues/498

I'll post links to the new install video and manual at issue 498, but worth noting is the updated network diagram for EMMA, that shows the network connection(s) all now go into the EMMA. i.e The sDongle is depreciated and the EMMA is 'THE' future standard device to deploy, that provides connectivity to the inverter(s) via RS485, if using a single WallCharger then it connects via Ethernet (that direct connects into the LAN port of the EMMA) and the EMMA then provides Local network access and upstream connectivity to the internet/FusionSolar via its WAN port.

**Comment by miguelcrego** [comment]
HI I read a post where it's possible to connect and read data from an elfin, without dongle. I installa and WLCRS detects the 2 inverters, but show data 5 seconds and stops 2 seconds... any idea?

**Comment by dorianhh** [comment]
Just wanted to add a comment, that now the cascaded inverters are found directly, without any proxy.
I added my 2 inverters + the smart meter, entering the IP address and slave ids: 1,2

**Reply by j1gg3** [reply]
Hey, can you explain a bit more about your setup?

I do have a 6KTL M1 which is the Master and in addition a 2KTL L1 which is connected via RS485. The discovery won't find the 2KTL. I checked within Fusion2000 App and ID of 2KTL is 2.

**Comment by dettofatto** [comment]
Two Inverters (10kT Master + 8kT Slave) - Read OK, Cannot Control Output Power

Hi all,
I have a setup with two Huawei inverters:

10kTL-M1 as master

8kTL-M1 as slave, connected via RS485

Here’s what I’ve tried:

Connecting to each inverter via its own Wi-Fi AP (I added two Wi-Fi dongles to Home Assistant):

I can read data from both inverters (using slave_id 1 and 2)

Communication is stable and fast

BUT: when I try to change the output power (e.g. derating), it has no effect on either inverter.

Connecting only to the master (slave ID 1):

Same result: I can read all the data (including slave inverter), but cannot change power output.

Connecting via the Huawei Smart Dongle:

I can see both inverters through slave IDs 1 and 2

However, the connection is very slow, and writing commands (like derating) still has no effect.

So far, I’ve been unable to control output power, no matter the method.

My questions:

Is it possible to independently control the power of each inverter in a cascaded configuration, or almost together?

Does derating work only when connected via RS485?

Has anyone successfully done power control (active power limit) with a master/slave Huawei setup?

Thanks in advance for your support!

**Reply by dettofatto** [reply]
https://github.com/dettofatto/huawei_sdongle_control/tree/main

I've used a Phyton script

**Reply by Roving-Ronin** [reply]
If your inverters are correctly installed you should have a single in the master inverter and it then wired via rs485 to the secondary inverter and smart meter (DTSU-666h for 3ph).

In this setup any grid export limit set will be applied to both inverters and seen as the total export limit for both inverters combined, with the master inverter then controlling the slave inverters output also (up to 3 inverters supported in a cascaded setup).

WLCRS integration is all that you need to do this.

**Reply by dettofatto** [reply]
My inverters are far from my home assistant and I woudlike monitor my inverters separatly (est and ovest), so i've connected 2 wifi to 2 inverter (now 3, 1 for dongle, 1 for master and one for slave) , But I can't write  with WLCRS (or fusion solar) 18000 (10+8 inverter) on my sun2000 10k inverter... Only 11000.
So...
